docs/zpure/index.md

```diff
@@ -0,0 +1,350 @@
+
+To motivate our example here let's define an `AccountState` type.
+
+```scala mdoc
+case class AccountState(balance: Int, open: Boolean)
+```
+
+We will also define a domain specific error type which we will use.
+
+```scala mdoc
+sealed trait AccountError
+
+case object InsufficientFunds extends AccountError
+```
+
+The most basic operators for working with state are `get` and `set`. The `get` operator retrieves the initial state and the `set` operator sets the updated state to the specified value.
+
+Let's use these operators to describe a computation that withdraws funds from the account.
+
+```scala mdoc
+def withdraw(amount: Int): ZPure[Nothing, AccountState, AccountState, Any, AccountError, Unit] =
+  for {
+    state <- ZPure.get[AccountState]
+    _     <- if (amount > state.balance) ZPure.fail(InsufficientFunds)
+             else ZPure.set(AccountState(state.balance - amount, state.open))
+  } yield ()
+```
+
+This example combines a couple of the features we have talked about so far.
+
+First, we are using a _for comprehension_ to chain together multiple computations
+
+Second, we are using both the error type and the state type here. If the customer has insufficient funds we fail immediately, otherwise we update the state.
+
+One important thing to note here is that since `ZPure` computations are never concurrent we don't need to worry about conflicting modifications to the state.
+
+For example, in the context of `ZIO` code analogous to the above would not be safe because the `get` and `set` would not be performed atomically so we would need to do the entire transaction in a single `modify` operation, returning a value indicating whether there were sufficient funds. In `ZPure` we don't have to do any of that so our code can be simpler.
+
+There are also `update` and `modify` operators on `ZPure` so for example we could factor out the logic of updating the account balance like this:
+
+```scala mdoc
+def decrementBalance(amount: Int): ZPure[Nothing, AccountState, AccountState, Any, Nothing, Unit] =
+  ZPure.update(state => AccountState(state.balance - amount, state.open))
+```
+
+If we are working with state then to actually run our computation we need to provide it with an initial state. The easiest way to do this is with the `provideState` operator, analogous to how we provide the environment.
+
+```scala mdoc:nest
+val withdrawalComputation: ZPure[Nothing, Any, AccountState, Any, AccountError, Unit] =
+  withdraw(10).provideState(AccountState(100, true))
+```
+
+We will use the `runEither` operator, which is a variant of `run` that allows the computation to fail and just converts the failure to the left side of an `Either`. The `run` and `runEither` operators just return the final value, so let's use the `get` operator to get the final state before we run our computation.
+
+```scala moc
+val updatedAccountState: Either[AccountError, AccountState] =
+  (withdrawalComputation *> ZPure.get).runEither
+```
+
```

Comment:
you could explain in here that we're "cleaning up data" in that we're not interested at all in the `Unit` value of `withdrawalComputation` all we want is resulting state whereas for `interestComputation` we were interested in a final double value. Do you follow me?

Reply:
Added something about this.